### PR TITLE
Update dependencies version range

### PR DESCRIPTION
**Pull request by spridev**
The hapi ecosystem has always been using `x.x.x` to declare dependencies version range.

This wasn't always the case in the past few months:

- [Update handlebars dependency version](https://github.com/hapijs/hapi/commit/4c52b531c4f127a1e193e61f1cd20da4f6aae6b2#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519)
- [Update most out of date dependencies](https://github.com/hapijs/hapi/commit/815c854dd29af68c322d350c3add6eedcaa08d99#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519)
- [Replace joi with validate](https://github.com/hapijs/hapi/commit/f46f864f7e38c3a64bdab93ab07d1ebac90a2aab#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519)
- [Update catbox](https://github.com/hapijs/hapi/commit/9abfd8670556992dc5403024bda3585294391690#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519)
- [Allow global multipart](https://github.com/hapijs/hapi/commit/85d7801cc8bc38c5ad30b5e29ed36e328617fb28#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519)

Running `npm update` or `npm install` uses `^` by default. This probably explains why this convention has not always been followed.

I think it makes sense to declare dependencies consistently. I tried running the tests using the lowest acceptable version for each package (even if they were deprecated) according to this pull request. I had to make three changes for the tests to pass:

- `@hapi/hoek` using the lowest non-deprecated version (9.0.4)
- `@hapi/podium` using the lowest non-deprecated version (4.1.0)
- `@hapi/validate` using the lowest non-deprecated version (1.1.1)

Here is the resulting `package.json`:

```json
{
  "name": "@hapi/hapi",
  "description": "HTTP Server framework",
  "homepage": "https://hapi.dev",
  "version": "20.1.0",
  "repository": "git://github.com/hapijs/hapi",
  "main": "lib/index.js",
  "engines": {
    "node": ">=12.0.0"
  },
  "files": [
    "lib"
  ],
  "keywords": [
    "framework",
    "http",
    "api",
    "web"
  ],
  "dependencies": {
    "@hapi/accept": "5.0.0",
    "@hapi/ammo": "5.0.0",
    "@hapi/boom": "9.0.0",
    "@hapi/bounce": "2.0.0",
    "@hapi/call": "8.0.0",
    "@hapi/catbox": "11.0.0",
    "@hapi/catbox-memory": "5.0.0",
    "@hapi/heavy": "7.0.0",
    "@hapi/hoek": "9.0.4",
    "@hapi/mimos": "5.0.0",
    "@hapi/podium": "4.1.0",
    "@hapi/shot": "5.0.0",
    "@hapi/somever": "3.0.0",
    "@hapi/statehood": "7.0.0",
    "@hapi/subtext": "7.0.0",
    "@hapi/teamwork": "5.0.0",
    "@hapi/topo": "5.0.0",
    "@hapi/validate": "1.1.1"
  },
  "devDependencies": {
    "@hapi/code": "8.0.0",
    "@hapi/inert": "6.0.0",
    "@hapi/joi-legacy-test": "npm:@hapi/joi@15.0.0",
    "@hapi/lab": "24.0.0",
    "@hapi/wreck": "17.0.0",
    "@hapi/vision": "6.0.0",
    "handlebars": "4.0.0",
    "joi": "17.1.1"
  },
  "scripts": {
    "test": "lab -a @hapi/code -t 100 -L -m 5000",
    "test-tap": "lab -a @hapi/code -r tap -o tests.tap -m 5000",
    "test-cov-html": "lab -a @hapi/code -r html -o coverage.html -m 5000"
  },
  "license": "BSD-3-Clause"
}
```

### CODE REVIEW

**Comment by devinivy** [comment]
Thanks for taking a careful look into this.  There are a couple things to be careful of with these types of changes:

1. Allowing versions of modules which have security vulnerabilities or bugs that affect hapi.  For example, here's a fix in statehood 7.0.2: https://github.com/hapijs/statehood/issues/71.  We want to ensure new installations of hapi would not allow the vulnerable version to be installed.
2. Allowing versions of modules which don't meet hapi's feature requirements.  If we need v4.1.0 of podium to use the `few()` method, then we don't want the version range in package.json to match v4.0.0.

I believe that for at least a couple years using the `^` format has been our preferred way of expressing version ranges in these types of situations.  For example, this commit from about a year ago shows an intentional update of several dependencies to address some bugs: https://github.com/hapijs/hapi/commit/85d7801cc8bc38c5ad30b5e29ed36e328617fb28

**Comment by spridev** [comment]
I initially thought security vulnerabilities wouldn't be a problem because once a security vulnerability is resolved, older versions are automatically marked as deprecated. Those deprecations are clearly shown to the user when installing a deprecated package through npm. I agree that it still makes sense to prevent vulnerable versions to be installed.

I will go through each package one by one and update the version range following your recommendations.

**Comment by spridev** [comment]
### Dependencies

- `@hapi/boom`: versions < 9.1.0 are deprecated.
- `@hapi/catbox`: version 11.1.0 isn't deprecated, has no security vulnerabilities and meets hapi's feature requirements.
- `@hapi/heavy`: version 7.0.0 isn't deprecated, has no security vulnerabilities and meets hapi's feature requirements.
- `@hapi/hoek`: versions < 9.0.4 are deprecated.
- `@hapi/podium`: version 4.1.0 isn't deprecated, has no security vulnerabilities and meets hapi's feature requirements.
- `@hapi/shot`: version 5.0.0 isn't deprecated, has no security vulnerabilities and meets hapi's feature requirements.
- `@hapi/statehood`: version 7.0.2 isn't deprecated, has no security vulnerabilities and meets hapi's feature requirements.
- `@hapi/teamwork`: versions < 5.1.0 are deprecated.
- `@hapi/validate`: versions < 1.1.1 are deprecated.

### Development dependencies

- `@hapi/inert`: version 6.0.1 isn't deprecated, has no security vulnerabilities and meets hapi's feature requirements.
- `@hapi/vision`: version 6.0.0 isn't deprecated, has no security vulnerabilities and meets hapi's feature requirements.
- `handlebars`: versions < 4.7.4 have security vulnerabilities ([optimist > minimist](https://www.npmjs.com/advisories/1179)). This issue has been fixed by [migrating from optimist to yargs](https://github.com/handlebars-lang/handlebars.js/commit/d79212a662d2aae106449338a28717352ec65ce6).

Here is the resulting `package.json` without any security vulnerabilities, deprecation notices or missing features (except for the legacy `@hapi/joi` version):

```json
{
  "name": "@hapi/hapi",
  "description": "HTTP Server framework",
  "homepage": "https://hapi.dev",
  "version": "20.1.0",
  "repository": "git://github.com/hapijs/hapi",
  "main": "lib/index.js",
  "engines": {
    "node": ">=12.0.0"
  },
  "files": [
    "lib"
  ],
  "keywords": [
    "framework",
    "http",
    "api",
    "web"
  ],
  "dependencies": {
    "@hapi/accept": "5.0.1",
    "@hapi/ammo": "5.0.1",
    "@hapi/boom": "9.1.0",
    "@hapi/bounce": "2.0.0",
    "@hapi/call": "8.0.0",
    "@hapi/catbox": "11.1.0",
    "@hapi/catbox-memory": "5.0.0",
    "@hapi/heavy": "7.0.0",
    "@hapi/hoek": "9.0.4",
    "@hapi/mimos": "5.0.0",
    "@hapi/podium": "4.1.0",
    "@hapi/shot": "5.0.0",
    "@hapi/somever": "3.0.0",
    "@hapi/statehood": "7.0.2",
    "@hapi/subtext": "7.0.3",
    "@hapi/teamwork": "5.1.0",
    "@hapi/topo": "5.0.0",
    "@hapi/validate": "1.1.1"
  },
  "devDependencies": {
    "@hapi/code": "8.0.0",
    "@hapi/inert": "6.0.1",
    "@hapi/joi-legacy-test": "npm:@hapi/joi@15.0.0",
    "@hapi/lab": "24.0.0",
    "@hapi/wreck": "17.0.0",
    "@hapi/vision": "6.0.0",
    "handlebars": "4.7.4",
    "joi": "17.1.1"
  },
  "scripts": {
    "test": "lab -a @hapi/code -t 100 -L -m 5000",
    "test-tap": "lab -a @hapi/code -r tap -o tests.tap -m 5000",
    "test-cov-html": "lab -a @hapi/code -r html -o coverage.html -m 5000"
  },
  "license": "BSD-3-Clause"
}
```

**Comment by Nargonath** [comment]
Thank you for looking into this carefully.

I'm worried changing the format from `x.x.x` to a pinned point version will require us to deploy a new version of hapi every time a fix or a minor lands on a dependency which IMO adds unnecessary burden. Any new installation will always have the latest dependencies with the current format without any action on our part regarding `hapi`. For existing project that's the developer responsibilities. I don't see much benefit from switching to this personally.

**Comment by spridev** [comment]
The `^` format indeed requires more frequent updates, but not for every fix/minor. Here's how it works:

- `^1.2.3` := `>=1.2.3 <2.0.0-0`
- `^0.2.3` := `>=0.2.3 <0.3.0-0`
- `^0.0.3` := `>=0.0.3 <0.0.4-0`

Basically we would use the `^` format whenever:

- a package is marked has deprecated/has a security vulnerability
- a package has a new feature required by hapi

as long as this is not a major update, which would fallback to the usual `x.x.x`.

Currently, we have a mix of `x.x.x` and `^` which isn't consistent. The `^` format has been first used in 85d7801cc8bc38c5ad30b5e29ed36e328617fb28. It made sense because those packages had security vulnerabilities, but they're also marked as deprecated on npm (which triggers a warning to the developer). Same story for 9abfd8670556992dc5403024bda3585294391690 and f46f864f7e38c3a64bdab93ab07d1ebac90a2aab.

On the other hand, I think 815c854dd29af68c322d350c3add6eedcaa08d99 is the result of `npm update` and should not have been merged.

If we follow @devinivy's recommendations, then this PR is good to go.
If we follow @Nargonath's recommendations, we should revert every `^` format to `x.x.x` and give the responsibility to the developer.

I'm waiting for your feedback before going any further.

**Comment by Nargonath** [comment]
Hum I'm hesitant on this one. I like that it is more consistent as you mentioned. On the other hand I wonder if the maintenance will not become cumbersome as a result. I didn't notice you also kept the `x.x.x` format. This [comment](https://github.com/hapijs/hapi/pull/4231) had me thought otherwise. If we land this we need to add a documentation somewhere for this:

>Basically we would use the ^ format whenever:
> 
> a package is marked has deprecated/has a security vulnerability
> a package has a new feature required by hapi

Probably here: https://github.com/hapijs/.github/blob/master/MAINTENANCE.md. I foresee this falling apart otherwise as time goes on especially considering multiple person can update deps.

**Comment by spridev** [comment]
The sample `package.json` in my comments are just here to demonstrate the lowest acceptable versions for each package because afaik there's no way to `npm install --lowest-version`. This is how I determined and tested which versions ranges could be allowed.

I kept the `x.x.x` format when there was no deprecation, security vulnerability or missing feature. Otherwise I used `^` followed by the lowest version meeting those criteria.

Yes it indeed makes sense to mention those convention somewhere once we end up with a convenient solution.

**Comment by Marsup** [comment]
Personally I'd go for consistency, and since the caret/tilde format is superior to the x format, I'd follow the npm guidelines everywhere. I'm wondering how it will affect the lockfiles of people if we do this on a minor/patch though. My 2¢.

**Comment by spridev** [comment]
I get your point. The entire hapi ecosystem is currently using `x.x.x`, which would involve reviewing and updating every module's `package.json`.

**Comment by Nargonath** [comment]
Perhaps @nlf if you have time, could you shed some light on this one please if you can: https://github.com/hapijs/hapi/pull/4231#issuecomment-782142768.

**Comment by Marsup** [comment]
I don't think we need absolutely every module to follow that pattern immediately, it can be a background task when there's some maintenance or future release happening on modules. And even if a year from now some modules are still not up-to-date, it wouldn't be that bad, would it?

**Comment by devinivy** [comment]
I think there might be some final tweaks to this PR, but I am inclined to accept it then peel off the broader conversation about changing our conventions to a separate issue.

File: package.json [hunk 1]
```diff
@@ -20,31 +20,31 @@
...
-    "@hapi/teamwork": "5.x.x",
+    "@hapi/teamwork": "^5.1.0",
     "@hapi/topo": "5.x.x",
-    "@hapi/validate": "^1.1.0"
+    "@hapi/validate": "^1.1.1"
   },
   "devDependencies": {
     "@hapi/code": "8.x.x",
-    "@hapi/inert": "^6.0.2",
+    "@hapi/inert": "^6.0.1",
```
Comment (devinivy) [inline package.json:42]: Let's leave this as is, as joi was dropped as a dependency from inert in 6.0.2.
```suggestion
    "@hapi/inert": "^6.0.2",
```

File: package.json [hunk 2]
```diff
@@ -20,31 +20,31 @@
...
     "@hapi/joi-legacy-test": "npm:@hapi/joi@15.x.x",
     "@hapi/lab": "24.x.x",
     "@hapi/wreck": "17.x.x",
-    "@hapi/vision": "^6.0.1",
-    "handlebars": "^4.5.3",
+    "@hapi/vision": "6.x.x",
```
Comment (devinivy) [inline package.json:46]: Another joi vs validate change 👍 
```suggestion
    "@hapi/vision": "^6.0.1",
```

File: package.json [hunk 3]
```diff
@@ -20,31 +20,31 @@
   "dependencies": {
     "@hapi/accept": "^5.0.1",
     "@hapi/ammo": "^5.0.1",
-    "@hapi/boom": "9.x.x",
+    "@hapi/boom": "^9.1.0",
     "@hapi/bounce": "2.x.x",
     "@hapi/call": "8.x.x",
-    "@hapi/catbox": "^11.1.1",
+    "@hapi/catbox": "^11.1.0",
```
Comment (devinivy) [inline package.json:26]:
```suggestion
    "@hapi/catbox": "^11.1.1",
```

File: package.json [hunk 4]
```diff
@@ -20,31 +20,31 @@
...
     "@hapi/catbox-memory": "5.x.x",
-    "@hapi/heavy": "^7.0.1",
-    "@hapi/hoek": "9.x.x",
+    "@hapi/heavy": "7.x.x",
```
Comment (devinivy) [inline package.json:28]:
```suggestion
    "@hapi/heavy": "^7.0.1",
```

File: package.json [hunk 5]
```diff
@@ -20,31 +20,31 @@
...
+    "@hapi/hoek": "^9.0.4",
     "@hapi/mimos": "5.x.x",
-    "@hapi/podium": "^4.1.1",
-    "@hapi/shot": "^5.0.1",
+    "@hapi/podium": "^4.1.0",
+    "@hapi/shot": "5.x.x",
```
Comment (devinivy) [inline package.json:32]:
```suggestion
    "@hapi/podium": "^4.1.1",
    "@hapi/shot": "^5.0.1",
```

File: package.json [hunk 6]
```diff
@@ -20,31 +20,31 @@
...
     "@hapi/somever": "3.x.x",
-    "@hapi/statehood": "^7.0.3",
+    "@hapi/statehood": "^7.0.2",
```
Comment (devinivy) [inline package.json:34]:
```suggestion
    "@hapi/statehood": "^7.0.3",
```

**Comment by spridev** [comment]
Here are the final changes:

- `@hapi/boom`: versions < 9.1.0 are deprecated.
- `@hapi/hoek`: versions < 9.0.4 are deprecated.
- `@hapi/teamwork`: versions < 5.1.0 are deprecated.
- `@hapi/validate`: versions < 1.1.1 are deprecated.
- `handlebars`: versions < 4.7.4 have security vulnerabilities ([optimist > minimist](https://www.npmjs.com/advisories/1179)). This issue has been fixed by [migrating from optimist to yargs](https://github.com/handlebars-lang/handlebars.js/commit/d79212a662d2aae106449338a28717352ec65ce6).

I'm surprised the checks didn't pass. I ran them in [my fork](https://github.com/spridev/hapi/tree/deps-ci) and everything went fine.

**Comment by nlf** [comment]
> Perhaps @nlf if you have time, could you shed some light on this one please if you can: [#4231 (comment)](https://github.com/hapijs/hapi/pull/4231#issuecomment-782142768).

it'll cause some package-lock changes for sure. in my experience, though, while annoying it's pretty expected when you update a dependency that some changes will end up in your package-lock and most people don't care. no different than if we bumped the required range on a dependency today

**Comment by nlf** [comment]
i am 100% in favor of changing our dep ranges to caret syntax. as mentioned by others, it's the default in npm, and it's the thing most folks are familiar with.

`1.x.x` and `^1.0.0` are functionally equivalent so this is purely cosmetic, but i do like the looks of it.

i agree also that it's not necessary to do everything at once, but it's also not harmful in any way if we were to update all of the ranges here all at once, and then other modules as we update them for other purposes we make the same changes there. it might give us less to keep track of since "make sure all ranges are caret style" is easier than "make sure all ranges for deps who use caret style deps are caret style"

**Review by Nargonath — APPROVED** [review]
The breaking tests seem to be due to flaky test since every tests pass now after you pushed again. LGTM

Thanks for the investigation work on this @spridev. 👍

**Comment by devinivy** [comment]
Thanks again @spridev!

**Pull request merged**